dev/PushNotifications/PushNotificationsLongRunningTask/NotificationListenerManager.cpp

```diff
@@ -15,30 +18,31 @@ void NotificationListenerManager::Initialize(std::shared_ptr<ForegroundSinkManag
 
 void NotificationListenerManager::AddListener(std::wstring appId, std::wstring processName)
 {
-    auto lock = m_lock.lock_exclusive();
-
     THROW_HR_IF(E_INVALIDARG, appId.empty());
     THROW_HR_IF(E_INVALIDARG, processName.empty());
 
-    if (m_notificationListeners.find(appId) == std::end(m_notificationListeners))
-    {
-        ComPtr<INotificationListener> listener;
-        THROW_IF_FAILED(MakeAndInitialize<NotificationListener>(&listener, m_foregroundSinkManager, appId, processName));
-        THROW_IF_FAILED(PushNotifications_RegisterFullTrustApplication(appId.c_str(), GUID_NULL));
-        THROW_IF_FAILED(PushNotifications_RegisterNotificationSinkForFullTrustApplication(appId.c_str(), listener.Get()));
-
-        AgileRef agileListener;
-        THROW_IF_FAILED(AsAgile(listener.Get(), &agileListener));
-        m_notificationListeners.insert({ appId, agileListener });
-    }
+    auto lock = m_lock.lock_exclusive();
+
+    // Make sure we keep the long running sink up-to-date with wpncore.
+    ComPtr<INotificationListener> newListener;
+    THROW_IF_FAILED(MakeAndInitialize<NotificationListener>(&newListener, m_foregroundSinkManager, appId, processName));
+    THROW_IF_FAILED(PushNotifications_RegisterNotificationSinkForFullTrustApplication(appId.c_str(), newListener.Get()));
+
+    // Before updating the listener in the map,
+    // remove any previous listener that belongs to the app.
+    m_notificationListeners.erase(appId);
```

Comment:
I'm not sure I understand this. Why update the map with a new Listener if one already exists? We only need one listner per app right?

Reply:
Please read the description. The faulty scenario is:

- App launches the first time
- App requests a channel
- App calls PushReceived -> The SDK calls the LRP to register a foreground sink (and a NotificationListener). We set a NotificationSink in wpncore.
- App closes
- App launches for the second time
- App requests a channel -> **Wpncore cleans the NotificationSink. Push notifications won't arrive to the LRP, and we have a stale Listener on the LRP.**

Reply:
Actually I think I can directly insert the new value without erasing first (a replacement). Let me look into this.

Reply:
I have some concerns with this fix. Before I voice them out though, can you please point me to where we clean the NotificationSink in wpncore as part of the last step?

Reply:
Note: We will have to slightly tweak our flow for LRP Sink registration if channel requests are delting the sink in wpncore.